Brands create/edit migration

zuk3975 opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | migrate brands edit/create actions to symfony. Introduced new TypedRegexConstraint, which replicates legacy constraints (IsGenericName, IsCatalogName etc.) regex patterns. Tiny mce editor is not currently available for multilang fields, so it should be implemented in another PR.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10543
| How to test?  | Create/Edit forms should work as in legacy. Slight improvements: Fixed hints of illegal characters in form. (In legacy some of illegal characters hints wasn't correct according to its validation rules).:warning: Assets may need to be rebuilt. Test using these routes: ` /admin-dev/index.php/sell/catalog/brands/new`  AND `/admin-dev/index.php/sell/catalog/brands/{brand_id}/edit`

Index action and rest of manufacturer actions in PR #9337 (until that after create/edit action user is redirected to add new brand page)

Also waiting for #12735 to be merged to finish https://github.com/PrestaShop/PrestaShop/pull/12795/files#diff-6f21bb46d5d2e6916d3a7e4c5938b447R109

Known issue: https://github.com/PrestaShop/PrestaShop/pull/12735#issuecomment-475202254

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12735)
<!-- Reviewable:end -->

zuk3975 committed:
Initial commit

zuk3975 committed:
Rendered manufacturer edit form

zuk3975 committed:
Implemented data provider for manufacturerForm and enabled javascript in manufacturer edit page

zuk3975 committed:
Manufacturer Edit action handler. [WIP]

zuk3975 committed:
[WIP] manufacturer create action

zuk3975 committed:
Logo image rendering on edit action

zuk3975 committed:
Implemented error messages in Manufacturer controller

zuk3975 committed:
Adjusted form behavior in multi/single shop mode

zuk3975 committed:
Enabled javascript for meta_keywords input

zuk3975 committed:
Added hints to add/edit form

zuk3975 committed:
[WIP] Validation for ManufacturersType

zuk3975 committed:
Minor adjustments in ManufacturerType

zuk3975 committed:
Introduced new LegacyConstraintValidator

zuk3975 commented on src/PrestaShopBundle/Validator/Constraint/LegacyConstraintValidator.php:
```diff
@@ -0,0 +1,104 @@
+
+/**
+ * Replicates legacy validation
+ */
```
@matks wdyt about this kind of legacy regex validation?

zuk3975 commented on src/PrestaShopBundle/Validator/Constraint/LegacyConstraintValidator.php:
```diff
@@ -0,0 +1,104 @@
+     * @param string $pattern
+     *
+     * @return string pattern
+     */
```
@matks and specially here. I replicated this method from legacy just in case, but in reality i think this constant is never defined. Guess it should be PREG_BAD_UTF8_ERROR at first place.

zuk3975 commented on src/PrestaShopBundle/Validator/Constraint/LegacyConstraintValidator.php:
```diff
@@ -0,0 +1,104 @@
+ * Replicates legacy validation
+ */
+class LegacyConstraintValidator extends ConstraintValidator
+{
```
@matks wdyt about this kind of legacy regex validation?

matks replied:
Legacy rather refers to code that is outdated as it does not fit the new needs. I'm not sure a regex can become legacy 😉 it can be added, improved/updated, deleted but not outdated.

So please dont name it Legacy 😛 we need to find a more suitable name that will last.

zuk3975 commented on src/PrestaShopBundle/Validator/Constraint/LegacyConstraintValidator.php:
```diff
@@ -0,0 +1,104 @@
+     * @return string pattern
+     */
+    private function cleanNonUnicodeSupport($pattern)
+    {
```
@matks and specially here. I replicated this method from legacy just in case, but in reality i think this constant is never defined. Guess it should be PREG_BAD_UTF8_ERROR at first place.

matks replied:
It's probably, but we cant be sure. It's all right: our job is here to rewrite the code following new methods and mechanisms while keeping the behavior "as is". So let's keep the regex right now. It might be re-evaluated and improved in a later work.

If we were 100% sure we could remove it, but with the doubt, we keep legacy behavior.

zuk3975 committed:
Added abstract getter of legacy manufacturer

zuk3975 committed:
Fixed image uploading and presentation in create/edit actions

zuk3975 committed:
Moved custom regex validation constraint ns

zuk3975 commented:
@matks addressed your comments

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php:
```diff
@@ -0,0 +1,183 @@
+     * Show & process manufacturer creation.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update'], request.get('_legacy_controller'))"
```
```suggestion
     *     "is_granted(['create'], request.get('_legacy_controller'))"
```

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php:
```diff
@@ -0,0 +1,183 @@
+                'The object cannot be loaded (or found)',
+                'Admin.Notifications.Error'
+            ),
+            MemoryLimitException::class => $this->trans(
```
I see here that you have tried to find all possible failures from the code being used in the controller Actions in order to translate them. Is it right ?

I never thought about that but it actually make sense 🤔 wdyt @sarjon @eternoendless ?

sarjon replied:
not sure, but i think i did that as well since hitting memory limit is "expected" result in some cases. As far as I remember, it comes from legacy image uploading.

zuk3975 replied:
Its for image uploading, I'm sure that user must be notified when trying to upload too big file or wrong format. Those exceptions comes from abstract classes that aren't in same domain level as i remember so I'm catching CoreException

matks commented:
@zuk3975 one minor mistake found 😄

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Manufacturer/ManufacturerType.php:
```diff
@@ -0,0 +1,173 @@
+                    ]),
+                ],
+            ])
+            //@todo: implement isCleanHtml constraint
```
I guess you mean https://github.com/PrestaShop/PrestaShop/pull/12163/files#diff-d7e5ef7648ac1e0921c3948466485f53R36 ?

matks commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/Blocks/form.html.twig:
```diff
@@ -0,0 +1,116 @@
+ *#}
+
+{% import '@PrestaShop/Admin/macros.html.twig' as ps %}
+
```
```suggestion
{% block manufacturer_form %}
```

with a `endblock` at the end of the page 😉 this will ease the job if someone needs to extend this form

zuk3975 committed:
Minor code optimization and fixes

zuk3975 commented:
@matks ready

matks commented:
@zuk3975 https://github.com/PrestaShop/PrestaShop/pull/12735 has been merged, which means `isCleanHtml` constraint can be used if you rebase on develop. Can you please perform one last rebase to fix this and then we go for QA ? 😉

matks commented:
@colinegin I think you can review wording for this PR 😉

colinegin commented:
@matks i've just reviewed the wording, just a question about this one : "The object cannot be loaded (or found)" ==> what is the object we are talking about ? Could it be a bit more specified ? What do you think ?

matks commented:
@colinegin This is a generic error message if at some point when processing a request, we cannot find the Brand (a.k.a manufacturer) that is the target of the request.
So for example it can happen if
- you try to display the EDIT page for a brand that does not exists
- you try to modify a brand, but it just got deleted (by someone else on the backoffice) so the modification cannot be performed because it is not available anymore

We can make it more specific "The brand cannot be loaded (or found)", but we used this wording here as it already exists in the catalog. "The object cannot be loaded (or found)" might need an addition to the translation catalog.

colinegin commented:
@matks alright, let's leave it this way then !

Thank you :)

zuk3975 committed:
Added legacy links to routing

zuk3975 committed:
Implemented CleanHtml constraint

zuk3975 commented:
> @zuk3975 #12735 has been merged, which means `isCleanHtml` constraint can be used if you rebase on develop. Can you please perform one last rebase to fix this and then we go for QA ?

@matk done

marionf commented:
@zuk3975 

1) When I click on Catalog > Brands & suppliers, I arrive on the page below instead of the brands listing

![capture d'écran_1178](https://user-images.githubusercontent.com/13449658/54627423-b4693900-4a73-11e9-8e43-71bf65c453cf.png)

2) Short description and description fields should have a tiny mce editor

3) Prohibited characters are not specified for Short description and description fields

4) When I click on Save I have the error below

![capture d'écran_1179](https://user-images.githubusercontent.com/13449658/54627603-0f9b2b80-4a74-11e9-8d04-cb085f49191e.png)

zuk3975 committed:
Fixed routing to suit unfinished page migration

zuk3975 commented:
> @zuk3975
> 
> 1. When I click on Catalog > Brands & suppliers, I arrive on the page below instead of the brands listing
> 
> ![capture d'écran_1178](https://user-images.githubusercontent.com/13449658/54627423-b4693900-4a73-11e9-8e43-71bf65c453cf.png)
> 
> 2. Short description and description fields should have a tiny mce editor
> 3. Prohibited characters are not specified for Short description and description fields
> 4. When I click on Save I have the error below
> 
> ![capture d'écran_1179](https://user-images.githubusercontent.com/13449658/54627603-0f9b2b80-4a74-11e9-8d04-cb085f49191e.png)

@marionf, sorry for that, i updated PR description now. So, 

2. Its a known issue that tiny mce is not yet implemented for multilang fields, so it should be done in another pull request.

3. Prohibited characters were shown wrong in legacy page - in reality those fields are only checked for valid html and does not prevent saving various symbols. So I just removed those misleading hints.

1. My mistake, I did routing a bit confusing until index action is not implemented(in #9337 PR). So fixed it now. After creating or editing, now redirects to new create page.

matks commented:
@zuk3975 https://github.com/PrestaShop/PrestaShop/pull/9337 is merged now 😉 this is going to make things easier

marionf commented:
@zuk3975 

> Its a known issue that tiny mce is not yet implemented for multilang fields, so it should be done in another pull request.

Humm, I think it has already been implemented on maintenance page ? @matks 

![capture d'écran_1184](https://user-images.githubusercontent.com/13449658/54679040-add7d180-4b06-11e9-9e76-2f4ef0b1d1c5.png)

1. There is a strange "B"

![capture d'écran_1182](https://user-images.githubusercontent.com/13449658/54677681-b67ad880-4b03-11e9-8a45-877c8dbb40ef.png)

@TristanLDD wdyt ?

2. When I edit a brand and click save or cancel I am redirected on the add brand page 
Maybe we need a rebase as #9337 has been merged ?

3. Could you resolve conflicts ?

zuk3975 committed:
Fixes after rebasing

zuk3975 commented:
@matks noticed some bug maybe regarding multi-shop.
When manufacturer is deleted, its only deleted in current shop context. Issues might appear, for example, its possible to reach one shop manufacturer for editing (by hard-coding url) from another shop context. Might be more issues with that...

zuk3975 committed:
Added missing form rest

zuk3975 committed:
Removed unnecessary input length coounter

matks commented:
> @matks noticed some bug maybe regarding multi-shop.
> When manufacturer is deleted, its only deleted in current shop context. Issues might appear, for example, its possible to reach one shop manufacturer for editing (by hard-coding url) from another shop context. Might be more issues with that...

⏩ this is an issue we might have with all identifiable objects actually (thanks @rokaszygmantas)
so we'll treat in in another topic, for all pages

matks commented:
@marionf From the material icon we have, I think the most relevant to replace that "B" (which is an icon) is a star. Wdyt ?

<img width="217" alt="Capture d’écran 2019-03-22 à 12 46 08" src="https://user-images.githubusercontent.com/3830050/54820768-8ae12500-4ca0-11e9-94d0-e1ac6ac9b203.png">

marionf commented:
@matks I am ok for the star, but let's wait for @TristanLDD answer :)

zuk3975 commented:
@matks rebased

matks commented:
@marionf We got a ✅for the star icon so I added it. Can you re-check ? With the exception of TinyMCE editor everything should be OK

marionf commented:
Yes, it's all good :+1:

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/Blocks/form.html.twig:
```diff
@@ -0,0 +1,119 @@
+      {% endset %}
+
+      {{ ps.form_group_row(manufacturerForm.short_description, {}, {
+        'label': 'Short description'|trans({}, 'Admin.Global'),
```
Please localize it in _Admin.Catalog.Feature_.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/Blocks/form.html.twig:
```diff
@@ -0,0 +1,119 @@
+      {% endif %}
+
+      {{ ps.form_group_row(manufacturerForm.meta_title, {}, {
+        'label': 'Meta title'|trans({}, 'Admin.Global'),
```
Please localize it in _Admin.Catalog.Feature_.

zuk3975 replied:
@LouiseBonnard what about other fields like Meta Description, description, meta keywords ?

LouiseBonnard replied:
@zuk3975, they are all properly localized in _Admin.Global_. ;-)

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/Blocks/form.html.twig:
```diff
@@ -0,0 +1,119 @@
+      }) }}
+
+      {% set keywordHint %}
+        {{ 'To add "tags," click inside the field, write something, and then press "Enter."'|trans({}, 'Admin.Catalog.Help') }}
```
Localization is fine here, but the source string needs to be reworked a bit, quotation marks are badly placed: _To add "tags", click inside the field, write something, and then press "Enter"._

LouiseBonnard replied:
@zuk3975, it has to be modified at the end of the string as well.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/edit.html.twig:
```diff
@@ -0,0 +1,43 @@
+ *#}
+
+{% set enableSidebar = true %}
+{% set layoutTitle = 'Edit: %name%'|trans({'%name%': manufacturerName}, 'Admin.Actions') %}
```
What is the string here, _Edit: %name%_?

matks replied:
It's in the "Edit a Brand page", if you want to edit "Adidas" brand then the displayed string is "Edit: Adidas"

zuk3975 committed:
Added validation in add/edit handlers

zuk3975 committed:
Changed some translation domains in form

matks commented:
Thank you @zuk3975 🎉 😄
